File: Leetcode/dfs_recursion_backtracking/1008_ConstructBinarySearchTreeFromPreorderTraversal/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def bstFromPreorder(self, preorder):
        """
        :type preorder: List[int]
        :rtype: TreeNode
        """
        if not preorder: return None
        if len(preorder)==1: return TreeNode(preorder[0])
        left, right = 1000, 1000
        for i, num in enumerate(preorder):
            if num < preorder[0]:
                left = i
                break
        for i, num in enumerate(preorder):
            if num > preorder[0]:
                right = i
                break
        root = TreeNode(preorder[0])
        root.left = self.bstFromPreorder(preorder[left:right])
        root.right = self.bstFromPreorder(preorder[right:])
        return root

File: Leetcode/dfs_recursion_backtracking/1008_ConstructBinarySearchTreeFromPreorderTraversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    /*/
    public TreeNode bstFromPreorder(int[] preorder) {
        return dfs(preorder, 0, preorder.length-1);
    }
        private TreeNode dfs(int[] preorder, int start, int end){
            if(start >  end) return null;
            if(start == end) return new TreeNode(preorder[start]);
            int left = end+1, right = end+1;
            boolean leftFound = false, rightFound = false;
            for(int i = start; i<=end; i++){
                if( !leftFound && preorder[i] < preorder[start] ){
                    left = i;
                    leftFound = true;
                }
                if( !rightFound && preorder[i] > preorder[start] ){
                    right = i;
                    rightFound = true;
                }
            }
            TreeNode root = new TreeNode(preorder[start]);
            root.left = dfs(preorder, left, right-1);
            root.right = dfs(preorder, right, end);
            return root;
        }/*/
    public TreeNode bstFromPreorder(int[] A) {
        int[] counter = {0};
        return bstFromPreorder(A, counter, Integer.MAX_VALUE);
    }

    public TreeNode bstFromPreorder(int[] A, int[] i, int bound) {
        if (i[0] == A.length || A[i[0]] > bound) return null;
        TreeNode root = new TreeNode(A[i[0]++]);
        root.left = bstFromPreorder(A, i, root.val);
        root.right = bstFromPreorder(A, i, bound);
        return root;
    }
}
